h2o-core/src/main/java/water/H2O.java

```diff
@@ -2137,11 +2136,11 @@ public static boolean checkUnsupportedJava() {
       return false;
     }
 
-    // Notes: 
-    // - make sure that the following whitelist is logically consistent with whitelist in R code - see function .h2o.check_java_version in connection.R
-    // - upgrade of the javassist library should be considered when adding support for a new java version
-    if (JAVA_VERSION.isKnown() && !isUserEnabledJavaVersion() && (JAVA_VERSION.getMajor()<8 || JAVA_VERSION.getMajor()>14)) {
-      System.err.println("Only Java 8, 9, 10, 11, 12 and 13 are supported, system version is " + System.getProperty("java.version"));
+    if (!Java.runningOnSupportedVersion()) {
```

Comment:
Isn't a less complex solution the better solution? One issue you can see that this list is not ordered. I would prefer simplicity as more lines of code -> more issues is usually the rule.

Reply:
True. But then the issue was there were numbers missing in the error message.

Reply:
I even have JUnit for that logic btw.

Reply:
The other approach is to have the production code simple and add a unit test that checks the expectations.

> I even have JUnit for that logic btw.

I see the unit test but it is still returning a Set.

At some point the message shouldn't be just listing all the Javas. We would want to say "Java 8 and later (up to Java 20) is supported" - eg. instead of all the Javas, or no?

Reply:
IDK. That's why I "designed" the holes in the range - unsupported versions in the range. Not sure if this will ever be used, but better safe than sorry. This Java class was supposed to be a generic solution. We could also manually list the versions - as we wish. This is internal to the class and might be changed at any time. 

Having it in the Java.java fike makes this logic testable. The final test a specific version works is the smoke test. But that won't test unallowed versions 

There seems to be too many problems at a time.

- Is it ok to keep it in the class ? To me yes. 
- If yes, would a guaranteed order in the very contract of the versions method be sufficient ? To me this should be done. 
- Callers of that class may decide to display any error message they want. These concerns are clearly separated. Nothing prevents the caller from verifying the versions actually form one or multiple intervals and print those in any desired way ?
- Should we start showing the interval range instead of full list now ? I think why not. 

Hope these questions are more specific to what we're talking about here.